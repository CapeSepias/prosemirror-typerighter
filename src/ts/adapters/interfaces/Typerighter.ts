export interface TypeRighterResponse {
  input: string;
  results: TypeRighterMatch[];
}

export interface TypeRighterMatch {
  fromPos: number;
  toPos: number;
  message: string;
  shortMessage: string;
  rule: TypeRighterRule;
  suggestedReplacements: string[];
}

export interface TypeRighterReplacement {
  value: string;
}

export interface TypeRighterType {
  typeName: string;
}

export interface TypeRighterRule {
  category: TypeRighterCategory;
  description: string;
  id: string;
  issueType: string;
}

export interface TypeRighterCategory {
  id: string;
  name: string;
}
