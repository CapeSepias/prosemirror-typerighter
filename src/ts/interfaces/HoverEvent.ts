import { ValidationOutput } from "./Validation";

interface HoverEvent {
  hoverLeft: number | undefined;
  hoverTop: number | undefined;
  validationOutput: ValidationOutput | undefined;
}

export default HoverEvent;
