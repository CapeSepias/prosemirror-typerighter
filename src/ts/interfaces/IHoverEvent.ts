import { IStateHoverInfo } from "../state/reducer";
import { IMatch } from "./IMatch";

interface IHoverEvent {
  hoverInfo: IStateHoverInfo | undefined;
  match: IMatch | undefined;
}

export default IHoverEvent;
