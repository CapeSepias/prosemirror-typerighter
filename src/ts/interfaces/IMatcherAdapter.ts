import {
  IBlock,
  IMatch,
  ICategory,
  IMatcherResponse,
  TMatchRequestErrorWithDefault
} from "./IMatch";

export interface IFetchMatchesOptions<TMatch extends IMatch = IMatch> {
  requestId: string;
  blocks: IBlock[];
  categoryIds: string[];
  onMatchesReceived: TMatchesReceivedCallback<TMatch>;
  onRequestError: TRequestErrorCallback;
  onRequestComplete?: TRequestCompleteCallback;
  documentUrl?: string;
}

/**
 * @internal
 */
export declare class IMatcherAdapter<TMatch extends IMatch = IMatch> {
  /**
   * Fetch the matches for the given inputs.
   */
  public fetchMatches: (opts: IFetchMatchesOptions<TMatch>) => void;

  /**
   * Fetch the currently available matcher categories.
   */
  public fetchCategories: () => Promise<ICategory[]>;

  constructor(apiUrl: string);
}

export type TMatchesReceivedCallback<TMatch extends IMatch = IMatch> = (
  response: IMatcherResponse<TMatch[]>
) => void;

export type TRequestErrorCallback = (
  matchRequestError: TMatchRequestErrorWithDefault
) => void;

export type TRequestCompleteCallback = (requestId: string) => void;
