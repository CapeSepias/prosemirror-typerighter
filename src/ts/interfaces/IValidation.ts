export interface IRange {
  from: number;
  to: number;
}

export interface IValidationInput {
  id: string;
  str: string;
  from: number;
  to: number;
}

export type IValidationOutput = IValidationInput & {
  annotation: string;
  suggestions?: string[];
  type: string;
};

export interface IValidationError {
  validationInput: IValidationInput;
  message: string;
}

export interface IValidationResponse {
  validationOutputs: IValidationOutput[];
  validationInput: IValidationInput;
}

export type IValidationLibrary = Array<
  Array<{
    regExp: RegExp;
    annotation: string;
    operation: "ANNOTATE" | "REPLACE";
    type: string;
  }>
>;

export const Operations: {
  [key: string]: "ANNOTATE" | "REPLACE";
} = {
  ANNOTATE: "ANNOTATE",
  REPLACE: "REPLACE"
};
