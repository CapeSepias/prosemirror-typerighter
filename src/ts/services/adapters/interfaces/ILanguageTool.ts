export interface ILTResponse {
  language: unknown;
  matches: ILTMatch[];
  software: unknown;
  warnings: unknown;
}

export interface ILTMatch {
  context: {
    text: string;
    offset: number;
    length: number;
  };
  length: number;
  message: string;
  offset: number;
  replacements: ILTReplacement[];
  rule: ILTRule;
  sentence: string;
  shortMessage: string;
  type: ILTType;
}

export interface ILTReplacement {
  value: string;
}

export interface ILTType {
  typeName: string;
}

export interface ILTRule {
  category: ILTCategory;
  description: string;
  id: string;
  issueType: string;
}

export interface ILTCategory {
  id: string;
  name: string;
}
