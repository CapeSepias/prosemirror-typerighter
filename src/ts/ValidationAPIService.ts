import {
  ValidationOutput,
  ValidationInput,
  ValidationError
} from "./interfaces/Validation";
import ValidationStateManager, {
  RunningServiceValidation
} from "./ValidationStateManager";
import IValidationService from "./interfaces/IValidationService";
import { LTResponse } from "./interfaces/LanguageTool";
import flatten from "lodash/flatten";

const serviceName = "[validationAPIService]";

export const ValidationEvents = {
  VALIDATION_SUCCESS: "VALIDATION_SUCCESS",
  VALIDATION_ERROR: "VALIDATION_ERROR"
};

/**
 * The validation service. Calls to validate() begin validations
 * for ranges, which are returned via an event.
 */
class ValidationService extends ValidationStateManager<RunningServiceValidation>
  implements IValidationService {
  constructor(private apiUrl: string) {
    super();
  }

  /**
   * Validate a Prosemirror node, restricting checks to ranges if they're supplied.
   */
  public async validate(inputs: ValidationInput[], id: string | number) {
    const results = await Promise.all(
      inputs.map(async input => {
        const body = new URLSearchParams();
        body.append(
          "data",
          JSON.stringify({
            annotation: [
              {
                text: input.str
              }
            ]
          })
        );
        body.append("language", "en-US");
        const validation = {
          id,
          validationInputs: inputs
        };
        this.addRunningValidation(validation);
        const response = await fetch(this.apiUrl, {
          method: "POST",
          headers: new Headers({
            "Content-Type": "x-www-form-urlencoded"
          }),
          body
        });
        if (response.status !== 200) {
          this.handleError(input, id, response.status, response.statusText);
          return [
            {
              validationInput: input,
              id,
              message: response.statusText,
              status: response.status
            }
          ];
        }
        const validationData: LTResponse = await response.json();
        const validationOutputs: ValidationOutput[] = validationData.matches.map(
          match => ({
            str: match.sentence,
            from: input.from + match.offset,
            to: input.from + match.offset + match.length,
            annotation: match.message,
            type: match.rule.description,
            suggestions: match.replacements.map(_ => _.value)
          })
        );
        this.handleCompleteValidation(id, validationOutputs);
        return validationOutputs;
      })
    );
    return flatten<ValidationError | ValidationOutput>(results);
  }

  /**
   * Cancel all running validations.
   */
  public cancelValidation = () => {
    this.cancelValidation();
  };

  /**
   * Handle an error.
   */
  handleError = (
    validationInput: ValidationInput,
    id: string | number,
    status: number,
    message: string
  ) => {
    this.emit(ValidationEvents.VALIDATION_ERROR, {
      validationInput,
      id,
      message,
      status
    } as ValidationError);
  };

  /**
   * Handle a completed validation.
   */
  private handleCompleteValidation = (
    id: string | number,
    validationOutputs: ValidationOutput[]
  ) => {
    const completeValidation = this.findRunningValidation(id);
    if (!completeValidation) {
      return console.warn(
        `${serviceName} Received validation from worker, but no match in running validations for id ${id}`
      );
    }
    this.emit(ValidationEvents.VALIDATION_SUCCESS, {
      id,
      validationOutputs
    });
    this.removeRunningValidation(completeValidation);
  };
}

export default ValidationService;
