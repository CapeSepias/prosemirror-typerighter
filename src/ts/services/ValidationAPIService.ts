import { IValidationInput } from "../interfaces/IValidation";
import { IValidationAPIAdapter } from "../interfaces/IValidationAPIAdapter";
import Store, { STORE_EVENT_NEW_VALIDATION } from "../store";
import { Commands } from "../commands";

/**
 * An example validation service. Calls to validate() begin validations
 * for ranges, configured via the supplied adapter. Validation results and
 * errors dispatch the appropriate Prosemirror commands.
 */
class ValidationService {
  constructor(
    private store: Store,
    private commands: Commands,
    private adapter: IValidationAPIAdapter
  ) {
    this.store.on(STORE_EVENT_NEW_VALIDATION, validationInFlight => {
      // If we have a new validation, send it to the validation service.
      this.validate(validationInFlight.validationInput);
    });
  }

  /**
   * Validate a Prosemirror node, restricting checks to ranges if they're supplied.
   */
  public async validate(validationInput: IValidationInput) {
    try {
      const validationOutputs = await this.adapter(validationInput);
      this.commands.applyValidationResult({
        validationOutputs,
        validationInput
      });
      return validationOutputs;
    } catch (e) {
      this.commands.applyValidationError({
        validationInput,
        message: e.message
      });
      return {
        validationInput,
        message: e.message
      };
    }
  }
}

export default ValidationService;
