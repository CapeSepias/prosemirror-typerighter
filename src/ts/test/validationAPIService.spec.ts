import fetchMock from "fetch-mock";
import { ILTReplacement } from "../adapters/interfaces/ILanguageTool";
import createLanguageToolAdapter from "../adapters/languageTool";
import { IValidationOutput } from "../interfaces/IValidation";
import ValidationAPIService, {
  ValidationEvents
} from "../services/ValidationAPIService";

const createResponse = (strs: string[]) => ({
  language: "",
  software: "",
  warnings: "",
  matches: strs.map(str => ({
    context: {
      text: str,
      offset: 0,
      length: str.length
    },
    length: str.length,
    message: "It's just a bunch of numbers, mate",
    offset: 0,
    replacements: [] as ILTReplacement[],
    rule: {
      category: {
        id: "numberCat",
        name: "The number category"
      },
      description: "Some type - use constants, jeez",
      id: "numbersID",
      issueType: "issueType"
    },
    sentence: str,
    shortMessage: "Bunch o' numbers",
    type: {
      typeName: "Some type - use constants, jeez"
    }
  }))
});

const createOutput = (str: string, offset: number = 0) =>
  ({
    id: "id",
    from: offset,
    to: offset + str.length,
    str,
    type: "issueType",
    suggestions: [],
    annotation: "It's just a bunch of numbers, mate"
  } as IValidationOutput);

const validationInputs = [
  {
    from: 0,
    to: 10,
    str: "1234567890"
  }
];

jest.mock("uuid/v4", () => () => "id");

describe("ValidationAPIService", () => {
  afterEach(fetchMock.reset);
  it("should issue a fetch given a validation input, resolving with a validation output and broadcasting the correct event", async () => {
    const service = new ValidationAPIService(
      createLanguageToolAdapter("endpoint/check")
    );
    fetchMock.mock("endpoint/check", createResponse(["1234567890"]));

    expect.assertions(2);

    service.on(ValidationEvents.VALIDATION_SUCCESS, _ => {
      expect(_).toEqual({
        id: "id",
        validationInput: validationInputs[0],
        validationOutputs: [createOutput("1234567890")]
      });
    });
    const output = await service.validate(
      validationInputs,
      "id"
    );
    expect(output).toEqual([createOutput("1234567890")]);
  });
  it("should handle multiple validation inputs", async () => {
    const service = new ValidationAPIService(
      createLanguageToolAdapter("endpoint/check")
    );
    const validationInputs = [
      {
        from: 0,
        to: 10,
        str: "1234567890"
      },
      {
        from: 20,
        to: 30,
        str: "1234567890"
      }
    ];
    fetchMock
      .once("endpoint/check", createResponse(["1234567890"]))
      .once("endpoint/check", createResponse(["1234567890"]), {
        overwriteRoutes: false
      });

    const output = await service.validate(
      validationInputs,
      "id"
    );
    expect(output).toEqual([
      createOutput("1234567890"),
      createOutput("1234567890", 20)
    ]);
  });
  it("should handle validation errors", async () => {
    const service = new ValidationAPIService(
      createLanguageToolAdapter("endpoint/check")
    );
    fetchMock.once("endpoint/check", 400);

    const output = await service.validate(
      [
        {
          from: 0,
          to: 10,
          str: "1234567890"
        }
      ],
      "id"
    );
    expect(output).toMatchSnapshot();
  });
  it("should handle requests with no inputs", async () => {
    const service = new ValidationAPIService(
      createLanguageToolAdapter("endpoint/check")
    );
    fetchMock.once("endpoint/check", 400);

    const output = await service.validate([], "id");
    expect(output).toEqual([]);
  });
});
