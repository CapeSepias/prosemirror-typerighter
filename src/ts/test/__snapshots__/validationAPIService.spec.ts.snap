// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationAPIService should handle validation errors 1`] = `
Object {
  "message": "Error fetching validations. The server responded with status code 400: Bad Request",
  "validationInput": Object {
    "from": 0,
    "id": "0-from:0-to:10",
    "str": "str",
    "to": 10,
  },
}
`;
